File: platforms/android/foundation/foundation/src/main/java/io/nativeblocks/foundation/modifier/Clickable.kt
package io.nativeblocks.foundation.modifier

import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.LocalIndication
import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalHapticFeedback
import io.nativeblocks.compiler.type.Modifier
import io.nativeblocks.compiler.type.ModifierData
import io.nativeblocks.compiler.type.ModifierEvent
import androidx.compose.ui.Modifier as UiModifier

@OptIn(ExperimentalFoundationApi::class)
@Modifier(
    keyType = "nativeblocks/clickable",
    name = "Clickable",
    description = "Makes the block clickable (and optionally long-clickable).",
    version = 1,
    versionName = "1",
)
@Composable
internal fun clickable(
    @ModifierData(description = "Whether click handling is enabled.", defaultValue = "true")
    enabled: Boolean = true,
    @ModifierData(description = "Whether a ripple indication is shown on click.", defaultValue = "true")
    ripple: Boolean = true,
    @ModifierData(description = "Whether a haptic feedback fires on long click.", defaultValue = "true")
    hapticsOnLongClick: Boolean = true,
    @ModifierEvent(description = "Triggered when the block is clicked.")
    onClick: (() -> Unit)? = null,
    @ModifierEvent(description = "Triggered when the block is long-clicked.")
    onLongClick: (() -> Unit)? = null,
): UiModifier {
    if (onClick == null && onLongClick == null) return UiModifier
    val haptic = LocalHapticFeedback.current
    val interactionSource = remember { MutableInteractionSource() }
    val indication = if (ripple) LocalIndication.current else null
    return UiModifier.combinedClickable(
        interactionSource = interactionSource,
        indication = indication,
        enabled = enabled,
        onLongClick = onLongClick?.let { callback ->
            {
                if (hapticsOnLongClick) haptic.performHapticFeedback(HapticFeedbackType.LongPress)
                callback()
            }
        },
    ) {
        onClick?.invoke()
    }
}

File: platforms/ios/foundation/Sources/NativeblocksFoundation/modifier/Clickable.swift
import NativeblocksCompiler
import NativeblocksRuntime
import SwiftUI

@Modifier(
    name: "Clickable",
    keyType: "nativeblocks/clickable",
    description: "Makes the block tappable (and optionally long-pressable).",
    version: 1,
    versionName: "1"
)
struct Clickable: ViewModifier {

    @ModifierData(description: "Whether tap handling is enabled.", defaultValue: "true")
    var enabled: Bool = true

    @ModifierData(description: "Whether a haptic feedback fires on long press.", defaultValue: "true")
    var hapticsOnLongClick: Bool = true

    @ModifierEvent(description: "Triggered when the block is tapped.")
    var onClick: (() -> Void)? = nil

    @ModifierEvent(description: "Triggered when the block is long-pressed.")
    var onLongClick: (() -> Void)? = nil

    @ViewBuilder
    func body(content: Content) -> some View {
        if enabled, onClick != nil || onLongClick != nil {
            let tappable = content
                .contentShape(Rectangle())
                .onTapGesture { onClick?() }
            if let onLongClick = onLongClick {
                tappable.onLongPressGesture {
                    if hapticsOnLongClick { performLongPressHaptics() }
                    onLongClick()
                }
            } else {
                tappable
            }
        } else {
            content
        }
    }
}

private func performLongPressHaptics() {
    #if os(iOS)
        UIImpactFeedbackGenerator(style: .medium).impactOccurred()
    #endif
}
